app/waf/src/controllers/key.controller.ts

```diff
@@ -0,0 +1,248 @@
+      ),
+    },
+  })
+  async find(
+    @param.query.object('filter', getFilterSchemaFor(Key))
+    filter?: Filter,
+  ): Promise<CollectionResponse> {
+    return new CollectionResponse(
+      Key,
+      await this.keyRepository.find({
+        where: {or: [{tenantId: await this.tenantId}]},
+      }),
+    );
+  }
+
+  @get(prefix + '/keys/{id}', {
+    responses: {
+      '200': Schema.response(Key, 'Successfully retrieve Key resource'),
+      '404': Schema.notFound('Can not find Key resource'),
+    },
+  })
+  async findById(
+    @param(Schema.pathParameter('id', 'Key resource ID')) id: string,
+  ): Promise<Response> {
+    return new Response(
+      Key,
+      await this.keyRepository.findById(id, {
+        where: {
+          and: [
+            {or: [{tenantId: await this.tenantId}]},
+            {id: id},
+          ],
+        },
+      }),
+    );
+  }
+
+  @post(prefix + '/keys/{id}/adcs/{adcId}', {
+    responses: {
+      '204': Schema.emptyResponse('Uploading Key resource'),
+      '404': Schema.notFound('Can not find trustedDeivceId'),
+      '422': Schema.unprocessableEntity(' Adc is not trusted'),
+    },
+  })
+  async keyUpload(
+    @param(Schema.pathParameter('id', 'Key resource ID')) id: string,
+    @param(Schema.pathParameter('adcId', 'ADC resource ID')) adcId: string,
+    @requestBody() keyObj: ContentBody,
+  ): Promise<void> {
+    let key: Key = await this.keyRepository.findById(id, {
+      where: {
+        and: [
+          {or: [{tenantId: await this.tenantId}]},
+          {id: id},
+        ],
+      },
+    });
+
+    let adc: Adc = await this.adcRepository.findById(adcId, undefined, {
+      tenantId: await this.tenantId,
+    });
+
+    if (!adc.management.trustedDeviceId) {
+      throw new HttpErrors.UnprocessableEntity(`Adc: ${adc.id} is not trusted`);
+    }
+    let keyContent = keyObj.content || null;
+    let name = keyObj.name || 'f5key';
+    let length = 0;
+
+    if (keyContent !== null) {
+      length = keyContent.length;
+    }
+
+    if (!keyContent || length === 0) {
+      throw new HttpErrors.UnprocessableEntity('empty key content');
+    }
+
+    try {
+      await this.asgMgr.uploadFile(
+        adc.management.trustedDeviceId!,
+        keyContent,
+        length,
+        name,
+      );
+
+      let remoteName = name + key.id;
+      let body = {
+        command: 'install',
+        name: `${remoteName}`,
+        'from-local-file': `/var/config/rest/downloads/${name}`,
+      };
+      //install the key
+      await this.asgMgr.icontrolPost(
+        adc.management.trustedDeviceId!,
+        '/mgmt/tm/sys/crypto/key',
+        body,
+      );
+
+      key.installed = true;
+      key.remotepath = `/Common/${remoteName}`;
+      await this.keyRepository.updateById(id, key);
+    } catch (error) {
+      throw new HttpErrors.UnprocessableEntity(
+        'upload key to asg service failed',
+      );
+    }
+  }
+
+  @get(prefix + '/keys/{id}/adcs/{adcId}', {
```

Comment:
This api should check the key file on BIG-IP, right?

Reply:
yes, and I have implemented this.